mixer/pkg/protobuf/yaml/dynamic/auth.go

```diff
@@ -0,0 +1,225 @@
+	}
+
+	// Load certificates, include public certs and certs from config.
+	caCertPool, err := loadCerts(mtlsCfg.GetCaCertificates())
+	if err != nil {
+		return nil, fmt.Errorf("failed build tls dial option: ca cert cannot be load %v", err)
+	}
+
+	tc := credentials.NewTLS(&tls.Config{
+		Certificates:       []tls.Certificate{peerCert},
+		RootCAs:            caCertPool,
+		InsecureSkipVerify: skipVerify,
+	})
+	return []grpc.DialOption{grpc.WithTransportCredentials(tc)}, nil
+}
+
+func buildOAuthToken(oauthCfg *policypb.OAuth) (oauth2.TokenSource, error) {
+	sp := oauthCfg.ClientSecret
+	if sp == "" {
+		return nil, errors.New("oauth secret cannot be empty")
+	}
+	secret, err := ioutil.ReadFile(sp)
+	if err != nil {
+		return nil, err
+	}
+	c := &clientcredentials.Config{
+		ClientID:     oauthCfg.ClientId,
+		ClientSecret: string(secret),
+		TokenURL:     oauthCfg.TokenUrl,
+		Scopes:       oauthCfg.Scopes,
+	}
+	return c.TokenSource(context.Background()), nil
+}
+
+func buildAuthToken(token []byte, tlsCfg *policypb.Tls) (*headerToken, error) {
+	if tlsCfg.TokenType == nil {
+		return nil, errors.New("token type should be specified")
+	}
+	at := &headerToken{
+		token: token,
+	}
+	switch tlsCfg.TokenType.(type) {
+	case *policypb.Tls_AuthHeader_:
+		ah := tlsCfg.GetAuthHeader()
+		at.header = "authorization"
+		if ah == policypb.PLAIN {
+			at.prefix = ""
+		} else if ah == policypb.BEARER {
+			at.prefix = "Bearer"
+		}
+	case *policypb.Tls_CustomHeader:
+		at.prefix = ""
+		at.header = tlsCfg.GetCustomHeader()
+	default:
+		return nil, errors.New("authToken type cannot be empty")
+	}
+	return at, nil
+}
+
+func buildPerRPCCredentials(tlsCfg *policypb.Tls) (credentials.PerRPCCredentials, error) {
+	switch tlsCfg.TokenSource.(type) {
+	case *policypb.Tls_Oauth:
+		ts, err := buildOAuthToken(tlsCfg.GetOauth())
+		if err != nil {
+			return nil, err
+		}
+		// nolint: govet
+		return oauth.TokenSource{ts}, nil
+	case *policypb.Tls_TokenPath:
+		// read token from given path
+		tp := tlsCfg.GetTokenPath()
+		if tp == "" {
+			return nil, errors.New("token path should not be empty")
+		}
+		token, err := ioutil.ReadFile(tp)
+		if err != nil {
+			return nil, err
+		}
+
+		// initialize per rpc credential based on token type
+		ht, err := buildAuthToken(token, tlsCfg)
+		return ht, err
+	default:
+		return nil, errors.New("unexpected tls token source type")
+	}
+}
+
+func buildTLSDialOption(tlsCfg *policypb.Tls, skipVerify bool) ([]grpc.DialOption, error) {
+	var opts []grpc.DialOption
+	// Load certificates, include system certs, istio certs and additional certs from config.
+	caCertPool, err := loadCerts(tlsCfg.GetCaCertificates())
+	if err != nil {
+		return nil, fmt.Errorf("failed build tls dial option: ca cert cannot be load %v", err)
+	}
+
+	tc := credentials.NewTLS(&tls.Config{
+		RootCAs:            caCertPool,
+		InsecureSkipVerify: skipVerify,
+	})
+	opts = append(opts, grpc.WithTransportCredentials(tc))
```

Comment:
Is there some expectation for the certification rotation here? Can we document that expired certs require re-connection?

Reply:
For regular TLS, root cert expiration is an issue.
For mTLS, root and private cert expiration are probably a more important issue.

Reply:
I have little context of this PR, but if the intention is to use Istio mTLS with this go code, I guess there can be some issue...

When Istio services are in `PERMISSVE` mode, the filter chain match is based on TLS ALPN = "istio", if not, we'll hand off to application container to terminate TLS, which will fail because it does not expect Istio's CA root cert. And PERMISSIVE mode, IIRC, is the default mode now.

/cc @mandarjog

Reply:
@kyessenov No there is no expectation for rotation, everything is file based. @incfly, correct me if I am wrong, istio ca cert expiry is one year, so I think it should probably be fine. And as always there will be no issue about cert rotation with mixer v2 :). 

@incfly We expect the adapter backend in mesh to load Istio CA certs and terminate TLS.

Reply:
adapater backend, do they have Istio sidecar? If not, how do they obtain Istio certificate then?

Reply:
@incfly No there will be no sidecar and it will load certs as other control planes. @kyessenov brought up that serverName should also be added into tls option, but unfortunately it is not yet in API. I will add that later so that PERMISSIVE would work.

Reply:
Is there a server-side example for OOP adapter using Istio CA certs with rotation?

Reply:
I don't think so, all control planes (galley, pilot, prom, except mixer since it uses sidecar for encrypted traffic) use static file based CA certs. We need a general way to operate this for istio components.